agent/config/file_watcher.go

```diff
@@ -0,0 +1,230 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-hclog"
+)
+
+const timeoutDuration = 200 * time.Millisecond
+
+type FileWatcher struct {
+	watcher          *fsnotify.Watcher
+	configFiles      map[string]*watchedFile
+	handleFunc       func(event *WatcherEvent) error
+	logger           hclog.Logger
+	reconcileTimeout time.Duration
+	cancel           context.CancelFunc
+	done             chan interface{}
+	toBeRemoved      chan string
+	toBeAdded        chan string
+}
+
+type watchedFile struct {
+	iNode   uint64
+	watched bool
+}
+
+type WatcherEvent struct {
+	Filename string
+}
+
+func New(handleFunc func(event *WatcherEvent) error) (*FileWatcher, error) {
+	ws, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	cfgFiles := make(map[string]*watchedFile)
+	w := &FileWatcher{watcher: ws, configFiles: cfgFiles, handleFunc: handleFunc, logger: hclog.New(&hclog.LoggerOptions{}), reconcileTimeout: timeoutDuration, done: make(chan interface{}), toBeAdded: make(chan string), toBeRemoved: make(chan string), cancel: cancelFunc}
+
+	go w.watch(ctx)
+	return w, nil
+}
+
+func (w *FileWatcher) add(filename string) error {
+	if err := w.watcher.Add(filename); err != nil {
+		return err
+	}
+	iNode, err := w.getFileId(filename)
+	if err != nil {
+		return err
+	}
+	w.configFiles[filename] = &watchedFile{iNode: iNode, watched: true}
+	return nil
+}
+
+func (w *FileWatcher) remove(filename string) error {
+	if err := w.watcher.Remove(filename); err != nil {
+		return err
+	}
+	delete(w.configFiles, filename)
+	return nil
+}
+func (w *FileWatcher) Remove(filename string) error {
+	timeout := time.After(timeoutDuration)
+	select {
+	case w.toBeRemoved <- filename:
+		return nil
+	case <-timeout:
+		return fmt.Errorf("file remove timedout %s", filename)
+	}
+}
+
+func (w *FileWatcher) Add(filename string) error {
```

Comment:
I thought about making the `Add` return a `chan error` and pass the channel to the main loop to return the error if any, I refrained because I don't think we need it in the implementation. I will add it if there is a need.